Additional filters

gyrus opened the pull request:
As discussed here: https://wordpress.org/support/topic/large-number-of-notifications-holding-up-ajax-response/ And here: https://wordpress.org/support/topic/working-on-changes-to-ass_group_notification_activity-function/

- I've altered the subscription options markup, and included the `bpges_wrap_input_labels` filter to control this markup (defaults to true).
- I've added the `bpges_email_from_name` and `bpges_email_from_email` filters to control the email from name / address.
- I've added the `bpges_notification_link` filter to control the link included in notifications.

Let me know if you've any questions.

boonebgorges commented:
Hi @gyrus - Thanks for the PR.

The new filters seem like no problem. I've merged them into the 4.0.x branch and made some minor changes to formatting and documentation. See the linked changesets above.

As for the 'bpges_wrap_input_labels' filter - this seems like a real pain. I can understand how the wrapped labels might cause styling headaches, so why don't we just move them out and skip the filter? Could you refactor your changes to do this? It's possible that we will need a bit of modified styling to ensure that the default appearance is the same after the change as it is before the change.

gyrus commented:
@boonebgorges Thanks! Happy to remove the wrap labels filter and not wrap them - but are you sure it won't break someone else's styling based on the wrapped version?

boonebgorges commented:
It might. I'd wager that 95% of installations are using the default styling, so if we update the shipped styles so that the interface looks the same, most people won't notice.

gyrus commented:
@boonebgorges Sorry for the delay in getting back. The latest commit here should have set the input / label layout to not wrap the label around the input. Looks fine with a vanilla Twenty-Twenty-One installation:

![Screenshot 2022-01-11 at 18 42 04](https://user-images.githubusercontent.com/609161/149003174-c18c1ed4-9141-4427-afe4-a3fdecdcf5b1.png)

Let me know when this makes it into core...

boonebgorges commented:
The changes you've made for the interface that you screenshotted (Group > Email Options) look good to me.

The changes for the Manage > Email Options page cause the label to appear on a different line from the radio button. I'm able to do a quick and dirty fix for this, but it does create inconsistent interfaces on this page, since BP's own radio button interface uses wrapping labels. See screenshot.

![Screenshot_2022-01-11_15-01-33](https://user-images.githubusercontent.com/246627/149021054-064d841b-73cd-4195-a6a9-b29ba127066c.png)

All things being equal, I'd like to be consistent with the way that BP works. What would you think about adopting your changes for Group > Email Options but not for Group > Manage > Email Options?

gyrus commented:
@boonebgorges On my vanilla installation, I'm not seeing this under Group > Manage > Email Options. I just see 'Send an email notice to everyone in the group' and 'Welcome Email' forms. Looks like your screenshot is from Group > Manage > Settings?

In any case, I see what you mean. I feel like we've immediately hit an instance where the original idea of a filter to control this markup (with the default as the current approach) would be justified.

boonebgorges commented:
> @boonebgorges On my vanilla installation, I'm not seeing this under Group > Manage > Email Options. I just see 'Send an email notice to everyone in the group' and 'Welcome Email' forms. Looks like your screenshot is from Group > Manage > Settings?

Yeah, I probably told you the wrong place.

> In any case, I see what you mean. I feel like we've immediately hit an instance where the original idea of a filter to control this markup (with the default as the current approach) would be justified.

I agree that we are quickly running into a case where we need flexibility, but I still feel uneasy about your proposal. We can't introduce individual filters for every little markup variation that people might want. (What if I wanted the labels to appear after the inputs in the markup, rather than before?) A more flexible approach would be to have the entire markup be flexible, either as a template that can be overridden in the theme, or by building markup that can then be filtered in its entirety. The first option seems more in keeping with what other WP plugins do (and what BPGES already does for parts of its interface - see the `templates` directory). What do you think?

gyrus commented:
@boonebgorges I didn't sense the filters as the start of a slippery slope of code inflation, but I'm biased and don't have to maintain the code! If it's between templating or filtering the markup... Templating is easier to work with, but there's the risk that changes in the core code will be lost if you don't keep your templates up-to-date (no one does). I guess you're best placed to know how likely important changes of this kind are, to inform a templating vs. filtering decision. I'm happy to contribute either way. Of course, if you decide on templating, whatever I do would need to be following from some basic decisions you make about how to handle templating.

boonebgorges commented:
> I didn't sense the filters as the start of a slippery slope of code inflation, but I'm biased and don't have to maintain the code! If it's between templating or filtering the markup.

Thank you for being empathetic :-D

Templating is probably the best route. Let's create a new directory `/templates/bpges` and put the new templates there. Then, we'll probably need to remove the `bp_is_group_admin_page()` check from `bpges_register_template_stack()`. Then, use `bp_get_template_part( 'bpges/subscription-options' )` to call up the template. Does this seem reasonable and doable?

gyrus commented:
@boonebgorges Sorry for the delay... I think I need to dive into coding to see how it goes. First off, I'm wondering about how to manage this. You last reply seems to suggest a single template. I guess both the instances of code we're looking at here, in `screen-admin.php` and `screen-notifications.php` are basically selecting from the same options, though there's plenty of variation and a single template would certainly need a few variables feeding into it. Is this the idea, or would you go for 2 templates for this? And would the template be the entire radio group, or a single radio option? I suspect the former, but it's not clear to me looking at the code.

boonebgorges commented:
Separate templates seem fine if the markup is different.

The template would be for the whole radio group. Basically, take the code that's currently generating the markup in the function, and dump it into a template file; then the function should load the template using `bp_get_template_part()`.

gyrus commented:
@boonebgorges Please check my latest commit. I've reverted all the changes I made, and put the subscription options markup into a new templating folder.

The one change I made to the default templates is I added an `id` attribute to the inputs, and a corresponding `for` attribute to the labels. This is necessary in my overrides, where I have the label _not_ wrapping the inputs. Not strictly necessary when they wrap, but I thought it was good practice, and makes it easier for people adapting the default templates to jig them around however they want.

There was one bit of non-markup code I had to carry over into `subscription-options-group.php`, for the group status. The local variable fell out of scope with the templating.

I've tested it on a vanilla installation, and with the project I use this on, with override templates in the custom theme. All looks good - let me know what you think.

boonebgorges commented on templates/bpges/subscription-options-group.php:
```diff
@@ -0,0 +1,38 @@
+?>
+
+<div class="ass-email-type" id="ass-email-type_no">
+	<label for="ass_group_subscribe_no"><input type="radio" name="ass_group_subscribe" id="ass_group_subscribe_no" value="no" <?php if ( $group_status == "no" || $group_status == "un" || ! $group_status ) { echo 'checked="checked"'; } ?>><?php _e( 'No Email????', 'buddypress-group-email-subscription' ); ?></label>
```
This is a lot of question marks on 'No Email' :) Can we take those out please?

boonebgorges reviewed (changes requested):
Thanks so much for working on this, and kudos for figuring out my meaning without my spelling it out too much :) The approach in the PR looks great. The addition of 'id' attributes is very good. Just one small change to request - see my comment about a changed string.

gyrus commented:
@boonebgorges Ah, sorry, that was from testing the templating system, to distinguish between defaults and overrides! I've removed the question marks.

boonebgorges commented:
Awesome! Let's do it.

gyrus commented:
@boonebgorges Hi there, any idea when 4.0.2 might be released? I'm looking to tie up this loose end by removing my forked version from my client's site, when the release comes out including my changes here.

boonebgorges commented:
Hello - Probably sometime in July.
